### ISSUE TITLE
Proposal: IE 11 Support

### ISSUE THREAD

**Issue by amirrustam** [cypress-io/cypress#6423]
> **Current Status** - _last updated February 11, 2020_
> 
> ✅ Under Consideration
> ❌ Not Prioritized
> ❌ Not Under Active Development
> ℹ️ _Cypress currently supports [Firefox](https://cypress.io/blog/2020/02/06/introducing-firefox-and-edge-support-in-cypress-4-0/) and Chrome-family browsers (including Edge and Electron)_.

**Objective**
---
Add support for running Cypress tests within the [IE 11](https://en.wikipedia.org/wiki/Internet_Explorer#Internet_Explorer_11) web browser.
> **IE 11 End of Life**
> According to Microsoft, development of new features for Internet Explorer has ceased. However, it will continue to be maintained as part of the support policy for the versions of Windows with which it is included.
> _Source: [Wikipedia](https://en.wikipedia.org/wiki/Internet_Explorer#End_of_life)_

> This is a **placeholder proposal issue**. Necessary details will be added **once this feature is prioritized**.

**Comment by kwiersma** [comment]
As someone who works on an application for the healthcare sector where IE 11 will be around a long time I would love to see support implemented. Kudos for considering this rather then just dropping plans because MS declared IE 11 EOL. :)

**Comment by jpike88** [comment]
There is still a massive amount of computers on IE 11 and while IT departments exist this will probably remain the case, so this would be very good to have.

**Comment by levenleven** [comment]
Hope Safari will get higher priority as it definitely not going to disappear and actually is new IE 😁

**Comment by jpike88** [comment]
@levenleven thats weird cause I never have Safari behaviour deviate from Chrome. shouldn’t IE11 be more important because of the much higher risk of deviation (such as use of polyfills)?

**Comment by levenleven** [comment]
@jpike88 jokes aside, both are important. I'm just hoping MS will try to kill IE faster because Edge should have compatability mode. Also more and more non enterprise projects just drop IE support

**Comment by Mouvedia** [comment]
There's already an issue for this: #1564
It has—as of today—96 👍 and a 15$ bounty; #6423 should be closed as a dupe of #1564.

**Comment by Mouvedia** [comment]
By closing the original issue you are losing the linked bounty.

**Comment by vladnauto** [comment]
We actually been using Cypress for more than a year and waiting for IE11 support, because we have a lot of customers in Japan, who are using IE11. Those clients are very important to us. So if Cypress doesn't plan to support IE11, we will have to switch to some other framework.

**Comment by tomkel** [comment]
@vladnauto Feel free to donate engineers or money to the effort

**Comment by juergba** [comment]
I'm a maintainer of Mocha, and sometimes I'm dreaming of a life without IE11.
If Mocha dropped/freezed IE11 support, what implications would it have for Cypress?

**Comment by lakshmanp21** [comment]
this is a must have IE11 support, as most of financial domains still have IE11 as primary browser

**Comment by morficus** [comment]
@juergba I don't think it's realistic for a project as large as Mocha to drop support for IE11 in the near term.
Microsoft's official statement on IE11 is that [it will continue to receive updates as long as the OS it is installed on is also still supported](https://support.microsoft.com/en-us/help/17454/lifecycle-faq-internet-explorer-and-edge). That means that as long as Windows 10 is still around then they will continue to support IE11.

It also goes with out saying that major corporations around the world have a heavy dependency on IE11. Edge failed to gain traction and "new Edge" is still extremely new. Mocha dropping support for IE11 in the near term would seriously hinder a lot of developers and projects.

**Comment by jpike88** [comment]
IE11 isn't going away for a long, long long time.

**Comment by jennifer-shehane** [comment]
Hey everyone, please try to keep the discussion focused on your personal or company needs for Cypress supporting IE11. We don't need a debate in general about IE11. We'd like to hear directly from people who use Cypress and want testing for the IE11 browser so that we can better evaluate our roadmap. 

@juergba Thanks for sharing. Evaluating our dependencies support for IE11 will definitely factor into our decision.

**Comment by nicooprat** [comment]
Hi, we work for big clients, mainly in french finance & insurance companies. Mixpanel stats show ~40% of IE 11 as of today, which is steady for the last few years. I suppose their IT stick on this browser as long as possible, and we will certainly see a big change when they change all their software at once ; but for now, it's the single most used browser by our customers. Second comes Chrome, but with a lot of different version, sometimes pretty outdated.

So, thanks for thinking about non-tech companies :) Looking forward to using Cypress with IE 11.

**Comment by estefafdez** [comment]
+1 to this idea, some of us are still supporting IE11 as browser and we need to test againt it, thanks!

**Comment by jtaylordigital** [comment]
We also support a lot of corporate clients that are still using IE11. The majority of our traffic comes from IE11, so Cypress adding support for it would be a big help.

**Comment by hukka** [comment]
In case it's easier than actual IE11 support, having Chromium Edge in the IE11 mode would already help.

**Comment by RicardoVaranda** [comment]
Hi @jennifer-shehane ,

Thanks for the clarification.

In our organisation (50k+ users) IE11 is still the default browser. Many legacy applications do not support Chrome yet. I have been trying to adopt within the organisation and everyone can clearly see the benefits of using it over other frameworks but at the end of the day if it's not browser agnostic it makes the conversation very difficult. I really hope you reconsider this and support IE11 in the near future or sadly we will have to look for an alternative solution.

**Comment by mali313** [comment]
My customer is using IE11 and their organization it is still heavily used

**Comment by admah** [comment]
Like others have said, as someone in the healthcare space, IE11 support is a must-have. We're working in large networks of hospitals that all require our application work in IE11.

**Comment by vidup** [comment]
Hi.
Here at Skillup we work in the french training space. We provide a SaaS to companies that help them involve their employees in their training process, as well as automate a lot of work for HR and managers.
We have different browser requirements depending on our users' scope: there are only a few HR users per company, so we officially support the recent versions of Chrome and Mozilla for their application. In this part we've started experimenting with Cypress and it's honestly been a great experience, so we will switch from Selenium soon. Congratulations on your work on Mozilla & Chromium Edge support, by the way !

On the other side, our clients' employees and managers often cannot chose their browser which is installed by default by their IT service.
This year, around 28% of our users are on IE11, and 27% for Edge (we hope most of them will update to the chromium based version sooner or later). It's been pretty stable for 3 years. Since they number between a few hundreds and a few thousands per company, we can't really do anything about it, and we're not big enough to reject potential clients because of their browser, so honestly even if it was 5% it would still be a prerequisite. This is what is preventing us from going full Cypress instead of our current Selenium E2E solution.
![image](https://user-images.githubusercontent.com/6061078/75801846-10f45f80-5d7c-11ea-899d-4bd65ae92347.png)

To be honest, I believe a lot of small to medium B2B software companies have the same constraint. 

Thanks again for your work. Cypress is a great product.

**Comment by lakshmanp21** [comment]
+1 Legacy systems as most of banking domains are still on IE11

**Comment by evillemez** [comment]
We're In exactly the same boat as @admah and others:

> Like others have said, as someone in the healthcare space, IE11 support is a must-have. We're working in large networks of hospitals that all require our application work in IE11.

Healthcare in particular is an issue and not going to change any time soon because we're dealing with large organizations and inflexible policies.

**Comment by anastasiaZen** [comment]
> We're In exactly the same boat as @admah and others:
> 
> > Like others have said, as someone in the healthcare space, IE11 support is a must-have. We're working in large networks of hospitals that all require our application work in IE11.
> 
> Healthcare in particular is an issue and not going to change any time soon because we're dealing with large organizations and inflexible policies.

The same thing in my company. We are really willing to keep using Cypress but there are concerns about IE11. It would be great if you make the decision to support it.

**Comment by mastacheata** [comment]
Has anyone actually ever tried what happens if you run Cypress in IE or EdgeHTML?
There's a bunch of comments in the driver referring to IE11, so it looks like some developers were preparing for this to happen in the future.

If there was a list of things broken/needed for IE11 and EdgeHTML it would at least be a start for developing that functionality.
Definitely willing to give some dev hours to this cause here.

**Comment by jennifer-shehane** [comment]
There was some initial proof of concept work started in this PR, which may give an idea to the extent of work required: https://github.com/cypress-io/cypress/pull/5224

**Comment by cnschwarz** [comment]
Most corporate clients I interacted with are using IE11 and the old Edge, unfortunately they will not be forced to replace these before Microsoft cuts off the support. Since the IT departments are usually enforcing IE11 and old Edge as the only browsers the user also has no other choice. So at least to have one of those two supported would be a great benefit and another reason to choose Cypress over another e2e test-framework. 

From personal experience in the old Edge we faced almost never an issue with javascript-based applications, it is most of the time IE11 which could have an issue. 
So if we have to make the case for using an e2e test framework clients will always want one which covers also the browsers they use, as for Cypress this is then not the case and they would vote for another option. 

To summarise my statement, I would vote to go with IE11 since the old Edge in my opinion 
 is much more stable and cross-browser testing is in my opinion for that one not crucial.

**Comment by tbredin** [comment]
+1 its the main browser we have that would benefit from test coverage.

Our biggest client is an old legal firm still using a lot of filing cabinets and fax machines, so getting off IE 11 ... let's just say its gonna be a huge difficulty for us for another decade at least

**Comment by arjunamrith** [comment]
Hey I really like cypress and would like to use it as our primary tool for e2e testing. We are currently on selenium and are evaluating cypress on the staging environments We now use cypress for chrome and firefox testing but 60% of our client base still uses IE11 hence we cannot fully commit to cypress. And I think this will be the case for the next 3-4 years at least hence we would greatly benefit from IE11 support. Keep up the good work, awesome job on this project so far 👏

**Comment by Vandivier** [comment]
Capital One CreditWise dev here.

Our current browser support policy indicates IE11 is officially supported as a Tier 2 app. My team specifically, and the other client-facing web projects I'm aware of, plan to support IE11 at a minimum through end of year.

**Comment by mattjanssen** [comment]
We develop a SaaS for enterprise companies and must support IE11 for the foreseeable future. We would love to use the paid Cypress solution if it supported IE11.

**Comment by poplevente** [comment]
We also would like to see IE11 support, as a lot of corporate clients are still using IE11 and probably will for quite some time

**Comment by mastacheata** [comment]
I guess everyone knows that IE11 is still in use at large corporate clients. That's the only kind of business where IE was widely used ever since IE6.

If IE11 support is wanted and you already work around the various IE problems, shave off an hour or two of your usual work and spend it on helping with the PR mentioned a couple of weeks ago by Jennifer Shehane.

IE is not everyones favorite browser, so if it's not one of us that needs IE support for business purposes, I sincerely doubt anyone else will jump on board and work on IE problems when there's still other stuff to do.

I'll probably have time to look into that on Thursday and will at least try to get some of the conflicts of that PR solved. Not sure how far I can get with my limited Open source budget (we get one day every month to work on stuff not directly related to our projects that might or might not benefit the company, including open source stuff that is not part of the actual project budget)

**Comment by AlexJeffcott** [comment]
We (AMBOSS gmbh which is a medical knowledge and learning platform) are switching to Cypress.

However, hospitals and physicians (especially in Germany) use ie11 heavily so this is a very serious drawback for us and our users.

**Comment by Factuursimpel** [comment]
+1 for this. It's the only thing Cypress is missing to be perfect.

**Comment by mwickett** [comment]
Although overall IE usage is relatively low as a percentage of traffic, at HashiCorp we generate more than enough leads from users on IE11 to justify supporting the experience, and therefore need to be able to test against it. We are a B2B company with many large enterprise clients, so not really a surprise there.

Completely understand that it's a massive effort to support it, but adding my +1 to it. 

Cypress is our only e2e test tool at the moment, and (sadly), we're going to have to spin Selenium to get some minimal test coverage for IE11. As many have pointed out, it's extremely brittle and we've been burned by IE being busted one too many times.

**Comment by TroyAlford** [comment]
I will +1 that my company produces software for the financial industry, and our experience is that 40+% of our site's clients are forced to use IE11 due to internal company policies. As a result, we have to support IE11.

**Comment by jpike88** [comment]
https://www.americaninno.com/atlanta/fundings/how-a-san-fran-vc-firm-grew-atls-cypress-io-to-help-developers-at-netflix-slack/


>Cypress.io, an Atlanta-based startup that offers automated testing for applications in web browsers, has completed a $9 million Series A round, led by previous investor Bessemer Venture Partners.

lol

**Comment by MichaelWMerritt** [comment]
It may seem redundant based on the other feedback here, but I want to +1 as this is huge blocker for most, if not all, companies that create applications for the insurance and financial industries in the US, and worldwide.  At last check, IE11 is the predominant browser of those institutions.  Having IE11 support would definitely increase adoption of Cypress!

**Comment by poplevente** [comment]
we basically had to give up on cypress because of the missing IE support :(

**Comment by Mouvedia** [comment]
As of today this issue has 154 👍  (tallying #1564). It's the 7th issue with the most +1, just behind support for Safari.

**Comment by darrendahl** [comment]
We use cypress for our full end to end tests but we have selenium just for a cross browser sanity test. Definitely would be nice to move completely to cypress and deprecate our selenium tests.

**Comment by crazyjaco** [comment]
We are a large regional news and media publisher. A large portion of our audience skews older using older windows machines that use IE or mobile traffic on Safari.

We do not currently use Cypress, but would like to. The IE limitation is the main thing holding us back.

**Comment by jennifer-shehane** [comment]
@mastacheata We want to hear people's feedback on their needs for IE 11 support, so we don't intend to lock the issue as long as the discussion is within that regard. 

You can unsubscribe from notifications directly from issue if you would like to no longer receive notifications for the thread.

**Comment by paulyoung5** [comment]
+1 - similar to other comments, our company provides software for other businesses across a variety of sectors, where there's constraints on browser usage etc. We've been road-testing Cypress for a handful of greenfield projects, it works really great! IE11 (and Safari) support would really help with advocating its use internally

I can imagine it's probably a huge task to support IE11, I don't know if it's got a DevTools API or anything. @jennifer-shehane is there a list of tasks for IE11 support anywhere, that people could look at + make contributions to help with the burden?

**Comment by chrisphillers** [comment]
Working in the healthcare system ie11 is very important to our product - and I cant see that ending anytime soon.

**Comment by vrodnyanskiy** [comment]
We would appreciate support for IE11 as it is a gap in our automation test coverage.

**Comment by Natesh11** [comment]
so, will there be a support for IE browsers in future? where do we stand on this as of today ? the answer to this is going to potentially decide our strategies and no doubt we all love cypress.

**Comment by raj-balaebail** [comment]
Adding to the chorus, working in FinTech a lot of clients still use IE. We are moving away from Selenium and evaluating Cypress or TestCafe for our needs. Cypress wins in almost every respect, but lack of IE support can be a potential deal breaker !!

**Comment by alexisgahon** [comment]
Working for an insurtech and our client and our support definitely need IE11 ... Cypress is by far the best tool for e2e testing ! But we have to use another one to test IE11

**Comment by patcon** [comment]
I contribute to pol.is, a platform that uses statistics and machine learning to run large scale consensus-building exercises. This system is part of the core of vTaiwan, a process by which Taiwan runs public consultations and crowdsources legislation.

https://www.technologyreview.com/2018/08/21/240284/the-simple-but-ingenious-system-taiwan-uses-to-crowdsource-its-laws/

Polis has adopted cypress as an e2e testing tool (yay for running tests on GitHub Actions!) But we've committed to supporting back to IE8 for the most widely-accessed "participation interface", so as not to disenfranchise citizens who must be able to contribute to public consultations. Would be rad if we could use cypress to get this assurance of support.

**Comment by patcon** [comment]
Also, might we be able to **split this issue into two: "request for comment" and "implemention"?**

I really feel for your encouragement of on-the-ground customer feedback, but can also see why many people would subscribe to this for quite different motives... Both are understandable ❤️

**Comment by gtrufitt** [comment]
At The Guardian we still support IE11 as a gracefully degraded browser which require JS to run as expected, so this would be good. We also believe that supporting IE11 is important as 11% of screen reader users still rely on it so our browser support numbers are not the the whole picture https://webaim.org/projects/screenreadersurvey8/#browsercombos

**Comment by Mouvedia** [comment]
#5224 needs to be reopened

**Comment by luciusyeah** [comment]
I agree with comments above: Support for IE 11 would very much be appreciated. Especially since [browserstack](https://www.browserstack.com/docs/automate/cypress) costs a small fortune.

**Comment by mjaeggi** [comment]
I think this should be considered before supporting IE11

https://www.theverge.com/2020/8/17/21372487/microsoft-internet-explorer-11-support-end-365-legacy-edge

**Comment by raminder** [comment]
I agree, there should be support of IE11 browser.I am working on creating new framework, we are choosing testcafe over cypress only because of IE11 support.

**Comment by IanVS** [comment]
@mjaeggi That article is a little confusing, but the key words are "across its Microsoft 365 apps and services next year."  Microsoft is not ending support for IE11 globally, and many people will continue to use it (sometimes through no choice of their own).

**Comment by begli80** [comment]
IE support is a must. In our company we decided to go with nightwatch because cypress doesn't support IE and Safari

**Comment by olliebreeden** [comment]
Not sure how feasible this would be, but Microsoft Edge has an IE mode. Does anyone have an idea of the effort required for cypress to enable this before a test?

**Comment by kostia7alania** [comment]
**+1000500** Legacy systems as most of banking domains are still on IE11

**Comment by greghqa** [comment]
Over a third of our customer base is on IE11, so we're having to consider Protractor or something else though we really like Cypress :(

**Comment by fearhq** [comment]
IE11 is still used by a large swath of tech-illiterate end-users that sort of just have it installed, and don't know any better. Unfortunately, we sometimes break IE11 functionnality, and require either labour-costly exploratory tests, or a second set of selenium based tests

**Comment by Tjerk-Haaye-Henricus** [comment]
IE11 Support would be awesome :)

**Comment by AlanKrygowski** [comment]
IE support would truly be awesome!

**Comment by kkoo95** [comment]
Running unit tests with Jest is (way) more featureful than using Karma/Jasmine, but its lack of great cross-browser support compared to Karma have always slowed me down switching to it. I think having Cypress supporting IE and Safari (in this order) would definitely be a game changer, because i could peacefully let Karma go 💔

**Comment by efreethy** [comment]
Would love to see this. My company has a significant number of clients still on IE11.

**Comment by Rbudhiraja** [comment]
Our projects are with the government of various countries and more than 8o% of them are on IE. We really want to switch to cypress and waiting for IE compatible version.

**Comment by JonasMS** [comment]
Hey, is there an update on this since Feb 2020?

My co is also considering migrating off of Cypress because we'll need IE11 support in the near-future.

**Comment by bahmutov** [comment]
There is some work we are doing on support Safari. There is no movement on supporting IE11 for now

**Comment by yaffol** [comment]
IE 11 support is still relevant for my organisation - we support modern chrome/firefox/edge and Safari version 9+ and IE 11.
The comment from @gtrufitt about screen readers is also very relevant.

**Comment by cuauhtlahuac** [comment]
> There is still a massive amount of computers on IE 11 and while IT departments exist this will probably remain the case, so this would be very good to have.

Where I work all departments still use IE for secure reasons

**Comment by davidcarboni** [comment]
IE 11 support is actually going to become _more_ important for us after end-of-life.

Because development will no longer naturally check that changes work in IE, we'll need to rely on something in CI to flag issues.

**Comment by an1meshk** [comment]
@jennifer-shehane Please update whether IE 11 got prioritized and any other details.

I'm starting to used Cypress in my React app and eagerly looking for more browser support, and IE 11 is of the most needed. As a major chunk of my customers are on IE11 :(

**Comment by cscheffauer** [comment]
+1 for IE11 support!

**Comment by Vondry** [comment]
I would also like to test in IE (at least IE 11).

**Comment by khotsoshomang** [comment]
IE11 testing is needed in my organization. We have been using Cypress for the past 2 years

**Comment by ladyflasheart** [comment]
The organisation I work for is currently looking at using Cypress for E2E testing, but lack of IE11 support is currently the biggest factor pushing us towards other options such as webdriver io or TestCafe. Our applications are primarily in the healthcare sector and many users are obliged to use IE11 still and this is unlikely to change any time soon.

**Comment by 06kellyjac** [comment]
I'm glad so many others are also expressing a need for this feature in but please just thumbs up the initial post.
There are a lot of people watching this issue and every time there's a +1 comment and or mentioning they need IE11 for customers, with no movement on the actual issue it gets a bit annoying.

Thanks

**Comment by sscotth** [comment]
https://github.com/cypress-io/cypress/issues/6423#issuecomment-590735156

> Hey everyone, please try to keep the discussion focused on your personal or company needs for Cypress supporting IE11. We don't need a debate in general about IE11. We'd like to hear directly from people who use Cypress and want testing for the IE11 browser so that we can better evaluate our roadmap.

**Comment by koenbouw** [comment]
+1 , and we still need to maintain it for a long time after EOL so this update would greatly improve our ability to catch issues. Thanks for all the great work you do!

**Comment by Laksh47** [comment]
+1, this would eliminate the need for manual testing just for IE11

https://blogs.windows.com/windowsexperience/2021/05/19/the-future-of-internet-explorer-on-windows-10-is-in-microsoft-edge/

**Comment by Pedro-Puga** [comment]
Are there any updates on the ie11 support or any workarounds?

**Comment by Mouvedia** [comment]
> Are there any updates on the ie11 support

@Pedro-Puga
It has been requested for [more than 3 years](https://github.com/cypress-io/cypress/issues/1564) and #5224 has been closed so don't hold your breath.
There's no reason to not prioritise a top 5 issue in terms of 👍 .

**Comment by LiteSun** [comment]
+1 for IE11 support

**Comment by philiptyler** [comment]
My company works in Healthcare with Epic and Cerner (Which run in 90%) of US hospitals. Both of these software host their apps in IE11. This unfortunate browser is not going away anytime soon

**Comment by Yaco-Lee** [comment]
+1 for IE11 support! I also work for a Healthcare company and most of our users still use IE. Thanks for considering this!

**Comment by Arpita-NDSC** [comment]
I also work for a Healthcare company and we have been thinking about using Cypress in our new project. But without the IE11 support, I cannot have a conversation about benefits of adopting this great tool! Please consider this request. Hoping to see the IE browser support soon. Thanks!

**Comment by VincentPuget** [comment]
+1 for IE11 support

**Comment by adamretter** [comment]
For our customers that support Windows users, not having support for IE 11 is a deal-breaker. We have used Cypress for many things in the past and enjoyed it, however for those customers we are having to identify a different solution as more than 50% of their users are still on IE11 sadly.

**Comment by sscotth** [comment]
@[tux-rampage](https://github.com/tux-rampage) That is not accurate. You are referring to support under mainline consumer versions of Windows.

https://techcommunity.microsoft.com/t5/windows-it-pro-blog/internet-explorer-11-desktop-app-retirement-faq/ba-p/2366549

Enterprise channels have another year (May 2023) and LTSC/Server go until Jan 2029.

**Comment by tux-rampage** [comment]
@sscotth Thanks for the clarification I was not aware of. So it seems we do not get rid of that Zombie-Browser 😞

**Comment by jennifer-shehane** [comment]
We're closing this issue since IE11 is retried and officially out of support as noted in this [blog post](https://blogs.windows.com/windowsexperience/2022/06/15/internet-explorer-11-has-retired-and-is-officially-out-of-support-what-you-need-to-know/#:~:text=After%2025%2B%20years%20of%20helping,today%2C%20June%2015%2C%202022.) from Microsoft.

If you have a case where you would like to support testing for IE Mode in Edge or some other case surrounding IE support, we suggest opening a new issue for tracking so that we can better prioritize that work separate from IE 11 support outright (which this issue was tracking).
